### DISCUSSION TITLE
Detect largest object and crop/trim to a 3:4 ratio and center the object into the canvas

### DISCUSSION THREAD

**Discussion by TNNSLAB** [ImageMagick/ImageMagick#4039] [Help] [answered]
I need to crop/trim this model a bit over his head top point and bottom most point and center it into a 3:4 canvas. 
Any help is appreciated. 

Original image : 
![t5](https://user-images.githubusercontent.com/46636486/128198580-c1fcc52d-4cd4-45d8-82ce-6a72cbf28db1.jpg)

Desired result : 
![desired](https://user-images.githubusercontent.com/46636486/128199256-cccc49ba-3d2d-4975-8dc5-53930e530717.jpg)

**Comment by snibgo** [comment]
What version of IM, on what platform?

The input image has a dark shadow in the left, at the edge of the white background paper. How do you want to process that? Perhaps we can always ignore the left 10% of the image. Perhaps also the right 10%.

The image is very noisy, perhaps dirt on the white paper. How do you want to process that? Perhaps ignore small non-white patches.

**Reply by TNNSLAB** [reply]
That was fast ! 

I was thinking to detect the biggest object/shape which is the model and crop around it, ignoring the left black edge.

So yeah, both left and right 10% could be cropped. Yes, ignoring the non-white patches would help a lot. 

@snibgo you can read minds !

**Reply by fmw42** [reply]
Perhaps could be done by thresholding or flood fill and then using connected components to find the largest bounding box of set of bounding boxes if large enough. The combine bounding boxes and resize to desired aspect. Then crop.

**Reply by TNNSLAB** [reply]
You guys are amazing ! If only i could implement this :)

**Comment by fmw42** [comment]
What version of ImageMagick and what platform/OS?

**Reply by TNNSLAB** [reply]
Mac OS Mojave
Version: ImageMagick 7.0.11-13 Q16 x86_64 2021-05-17

**Comment by snibgo** [comment]
Trimming down to the figure is fairly simple. Windows BAT syntax:
```
magick ^
  128198580-c1fcc52d-4cd4-45d8-82ce-6a72cbf28db1.jpg ^
  -set option:WW %%[fx:w*0.8] ^
  -set option:DX %%[fx:w*0.1] ^
  -set option:HH %%[fx:h] ^
  -set option:DY 0 ^
  -crop %%[WW]x%%[HH]+%%[DX]+%%[DY] ^
  -blur 0x10 ^
  -fuzz 5%% ^
  -trim ^
  -trim ^
  -trim ^
  x.jpg ^
```
Note that I trim three times.

**Comment by fmw42** [comment]
Here is a bash shell script that you can use on Mac in a terminal window. I shave 5% on the left and right first. Then threshold. Then get the largest bounding box after throwing out any region with area < 5000. Then adjust the dimensions for a crop to the desired aspect ratio. You must use 2:3 if vertical and 3:2 if horizontal. You could modify it to use 3:2 and then invert if H>W. Adjust the area threshold as desired and shave of 5% as desired.



![standing_man](https://user-images.githubusercontent.com/13458218/128219521-73fee8c4-95ad-4780-9d4a-1270a74a7688.jpg)



```
aspect="2:3"
bbox=`magick \
standing_man.jpg \
-shave 5x0% +repage \
+write tmp.png \
-auto-threshold otsu \
-negate \
-type bilevel \
-define connected-components:exclude-header=true \
-define connected-components:area-threshold=5000 \
-define connected-components:keep-top=1 \
-define connected-components:verbose=true \
-define connected-components:mean-color=true \
-connected-components 8 null: | \
grep "gray(255)" | awk '{print $2}' | tr "x" "+"`
echo "$bbox"
ww=`echo $bbox | cut -d+ -f1`
hh=`echo $bbox | cut -d+ -f2`
xo=`echo $bbox | cut -d+ -f3`
yo=`echo $bbox | cut -d+ -f4`
cx=`magick xc: -format "%[fx:$xo+$ww/2]" info:`
cy=`magick xc: -format "%[fx:$yo+$hh/2]" info:`
aw=`echo $aspect | cut -d: -f1`
ah=`echo $aspect | cut -d: -f2`
aspectwh=`magick xc: -format "%[fx:$aw/$ah]" info:`
aspect=`magick xc: -format "%[fx:$ww/$hh]" info:`
echo "ww=$ww; hh=$hh; xo=$xo; yo=$yo; cx=$cx; cy=$cy; aw=$aw; ah=$ah; aspectwh=$aspectwh; aspect=$aspect;"
test=`magick xc: -format "%[fx:$aspectwh>($ww/$hh)?1:0]" info:`
if [ $test -eq 1 ]; then
	wd=`magick xc: -format "%[fx:$hh*$aspectwh]" info:`
	xx=`magick xc: -format "%[fx:$cx-$wd/2]" info:`
	cropvals="${wd}x${hh}+${xx}+${yo}"
	echo "$cropvals;"
else
	ht=`magick xc: -format "%[fx:$ww/$aspectwh]" info:`
	yy=`magick xc: -format "%[fx:$cy-$ht/2]" info:`
	cropvals="${ww}x${ht}+${xo}+${yy}"
	echo "$cropvals;"
fi
magick tmp.png -crop $cropvals +repage standing_man_crop.jpg
rm -f tmp.png

```



![standing_man_crop](https://user-images.githubusercontent.com/13458218/128219475-17272078-d070-408c-8816-307f186f9df2.jpg)

**Comment by snibgo** [comment]
In a single "magick" command (Windows BAT syntax), assuming the figure is always too tall so we need to add to the width:
```
magick ^
  128198580-c1fcc52d-4cd4-45d8-82ce-6a72cbf28db1.jpg ^
  ( +clone ^
    -set option:WW %%[fx:w*0.8] ^
    -set option:DX %%[fx:w*0.1] ^
    -set option:HH %%[fx:h] ^
    -set option:DY 0 ^
    -crop %%[WW]x%%[HH]+%%[DX]+%%[DY] ^
    -blur 0x10 ^
    -fuzz 10%% ^
    -trim ^
    -trim ^
    -trim ^
    -set option:OW %%[fx:h*3/4] ^
    -set option:OH %%[fx:h] ^
    -set option:OX %%[fx:(page.width-h*3/4)/2] ^
    -set option:OY %%[fx:page.y] ^
    +delete ^
  ) ^
  -crop %%[OW]x%%[OH]+%%[OX]+%%[OY] +repage ^
  x2.jpg
```
For bash syntax, change line ends from ^ to \\, escape the parentheses \\( and \\), and change double %% to single %.
![x2](https://user-images.githubusercontent.com/33812368/128283333-e07e1dbb-25f8-4462-b824-ae2cd03f99c0.jpg)

**Reply by TNNSLAB** [reply]
hey Snibo, i converted it to bash, but getting some errors :( 

magick \
  t5.jpg \
  ( +clone \
    -set option:WW %[fx:w*0.8] \
    -set option:DX %[fx:w*0.1] \
    -set option:HH %[fx:h] \
    -set option:DY 0 \
    -crop %[WW]x%[HH]+%[DX]+%[DY] \
    -blur 0x10 \
    -fuzz 10% \
    -trim \
    -trim \
    -trim \
    -set option:OW %[fx:h*3/4] \
    -set option:OH %[fx:h] \
    -set option:OX %[fx:(page.width-h*3/4)/2] \
    -set option:OY %[fx:page.y] \
    +delete \
  ) \
  -crop %[OW]x%[OH]+%[OX]+%[OY] +repage \
  x2.jpg
  
  I'm getting this : 

-bash: syntax error near unexpected token `('
Imaging-Adrian-2:Desktop imaging-adrian$     -set option:OY %[fx:page.y] \
>     +delete \
>   ) \
-bash: syntax error near unexpected token `)'
Imaging-Adrian-2:Desktop imaging-adrian$   -crop %[OW]x%[OH]+%[OX]+%[OY] +repage \
>   x2.jpg
-bash: -crop: command not found

any ideas ?

**Comment by snibgo** [comment, marked as answer]
When pasting code into the forum, please use the markup tools provided. Or use a line with just three backticks at the start and at the end of the code. Otherwise, characters like asterisk \* and backslash \\ get stripped, and posts are difficult to understand.

As I said, for bash you need to escape the parentheses, and perhaps you didn't. This means putting a backslash in front of each one, like this: \\( and \\).

EDIT to add: there is a parenthesis pair in a %[fx:] expression, so those also need to be escaped, or quoted like this:
```
magick \
  128198580-c1fcc52d-4cd4-45d8-82ce-6a72cbf28db1.jpg \
  \( +clone \
    -set option:WW %[fx:w*0.8] \
    -set option:DX %[fx:w*0.1] \
    -set option:HH %[fx:h] \
    -set option:DY 0 \
    -crop %[WW]x%[HH]+%[DX]+%[DY] \
    -blur 0x10 \
    -fuzz 10% \
    -trim \
    -trim \
    -trim \
    -set option:OW "%[fx:h*3/4]" \
    -set option:OH "%[fx:h]" \
    -set option:OX "%[fx:(page.width-h*3/4)/2]" \
    -set option:OY "%[fx:page.y]" \
    +delete \
  \) \
  -crop %[OW]x%[OH]+%[OX]+%[OY] +repage \
  x2.jpg
```

**Reply by TNNSLAB** [reply]
@snibgo thank you very much ! can I adjust somehow the top/bottom cropping values ? I want to leave a little bit more space both on top and on the bottom. 
Thank you and sorry for late super later reply

**Comment by snibgo** [comment]
You can adjust the expressions for OW, etc. For example, to leave 30 pixels more at the top and 20 at the bottom, so the height increases by 50 pixels, and the width increases by 3/4 of that:
```
    -set option:OW "%[fx:(h+50)*3/4]" \
    -set option:OH "%[fx:h+50]" \
    -set option:OX "%[fx:(page.width-h*3/4)/2]" \
    -set option:OY "%[fx:page.y-30]" \
```

**Reply by TNNSLAB** [reply]
Nice ! 
It's working amazing when I do it on just a single file. 
As soon as I do multiple, the results are not good. Any ideas why ? What am I doing wrong ? 
Basically same file that i'm running single it's perfect and when in batch it crops into the person. 


```
magick \
  test/*.jpg \
  \( +clone \
    -set option:WW %[fx:w*0.8] \
    -set option:DX %[fx:w*0.1] \
    -set option:HH %[fx:h] \
    -set option:DY 0 \
    -crop %[WW]x%[HH]+%[DX]+%[DY] \
    -blur 0x10 \
    -fuzz 10% \
    -trim \
    -trim \
    -trim \
    -set option:OW "%[fx:(h+160)*3/4]" \
    -set option:OH "%[fx:h+160]" \
    -set option:OX "%[fx:(page.width-h*3/4)/2]" \
    -set option:OY "%[fx:page.y-70]" \
    +delete \
  \) \
  -crop %[OW]x%[OH]+%[OX]+%[OY] +repage \
  test_out/*.jpg

```

**Reply by fmw42** [reply]
Bash/ImageMagick does not support wild cards for both input and output. You will need to write a script "for" loop over your directory or list of images. You can however, use a wild card for the input and test_out.jpg will result in test_output-0.jpg, test_output-1.jpg etc, that is, same output name with automatic numbering.

**Reply by TNNSLAB** [reply]
@fmw42 thank you for your detailed answer. i'm trying to figure that loop part...also...can I convert the script above into a single line ? I want to use it later in Python...is this an option or the structure is important in order for the script to work. 
P.S. : sorry if the questions are too dumb...

**Reply by fmw42** [reply]
The code you have is only one command line broken into multiple line just for viewing. It will run fine in a loop.

```
cd to input directory
for img in *.jpg; do
inname=$(magick "$img" -format "%t" info:)
magick \
  "$img" \
  \( +clone \
    -set option:WW %[fx:w*0.8] \
    -set option:DX %[fx:w*0.1] \
    -set option:HH %[fx:h] \
    -set option:DY 0 \
    -crop %[WW]x%[HH]+%[DX]+%[DY] \
    -blur 0x10 \
    -fuzz 10% \
    -trim \
    -trim \
    -trim \
    -set option:OW "%[fx:(h+160)*3/4]" \
    -set option:OH "%[fx:h+160]" \
    -set option:OX "%[fx:(page.width-h*3/4)/2]" \
    -set option:OY "%[fx:page.y-70]" \
    +delete \
  \) \
  -crop %[OW]x%[OH]+%[OX]+%[OY] +repage \
  "test_out/${inname}.jpg"
done
```

For Python you would have to use a subprocess call to run a single ImageMagick command line and do the "for" looping in Python over your directory for input and output.

Or you would have to do all your processing in Python/OpenCV/Numpy. But that is going to be harder.

**Reply by TNNSLAB** [reply]
Fantastic ! It works ! You guys help us all so much ! So detailed ! I see your nickname on so many places ! SO, Reddit, here...unbelievable ! Thank you guys for your time !

**Reply by TNNSLAB** [reply]
@fmw42 One last question please... 

I tried to convert your code into one command line and i can't make it to work. The syntax is not correct for all the ``` '-set option:WW %[fx:w*0.8]' ```
This is the way I'm running Imagemagick in Python and it's working: 
```
os.system(convert test.jpg -geometry 1800x2400^ -gravity center -crop 1800x2400+0+0  test_out.jpg")
```

**Reply by fmw42** [reply]
To make the command into one line, just remove the line ending \ and collapse all the code to one long line of text.

**Comment by snibgo** [comment]
@schwarzwals: Have you consulted Python documentation? "%" is a special character in Python, so the Python interpreter would remove it. You need it to pass the "%" to IM, so you need to escape it in the Python command.
